modules/ml/src/main/java/org/apache/ignite/ml/recommendation/RecommendationTrainer.java

```diff
@@ -0,0 +1,319 @@
+import org.apache.ignite.ml.dataset.primitive.builder.context.EmptyContextBuilder;
+import org.apache.ignite.ml.dataset.primitive.context.EmptyContext;
+import org.apache.ignite.ml.environment.LearningEnvironment;
+import org.apache.ignite.ml.environment.LearningEnvironmentBuilder;
+import org.apache.ignite.ml.math.primitives.vector.Vector;
+import org.apache.ignite.ml.math.primitives.vector.VectorUtils;
+import org.apache.ignite.ml.recommendation.util.MatrixFactorizationGradient;
+import org.apache.ignite.ml.recommendation.util.RecommendationDatasetData;
+import org.apache.ignite.ml.recommendation.util.RecommendationDatasetDataBuilder;
+
+/**
+ * Trainer of the recommendation system.
+ */
+public class RecommendationTrainer {
+    /** Environment builder. */
+    private LearningEnvironmentBuilder environmentBuilder = LearningEnvironmentBuilder.defaultBuilder();
+
+    /** Trainer learning environment. */
+    private LearningEnvironment trainerEnvironment = environmentBuilder.buildForTrainer();
+
+    /** Batch size of stochastic gradient descent. The size of a dataset used on each step of SGD. */
+    private int batchSize = 1000;
+
+    /** Regularizer. */
+    private double regularizer = 0.0;
+
+    /** Learning rate. */
+    private double learningRate = 10.0;
+
+    /** Max number of SGD iterations. */
+    private double maxIterations = 1000;
+
+    /** Number of rows/cols in matrices after factorization. */
+    private int k = 10;
+
+    /**
+     * Fits prediction model.
+     *
+     * @param datasetBuilder Dataset builder.
+     * @param <K> Type of a key in {@code upstream} data.
+     * @param <O> Type of an object.
+     * @param <S> Type of a subject.
+     * @return Trained recommendation model.
+     */
+    public <K, O extends Serializable, S extends Serializable> RecommendationModel<O, S> fit(
+        DatasetBuilder<K, ? extends ObjectSubjectRatingTriplet<O, S>> datasetBuilder) {
+        try (Dataset<EmptyContext, RecommendationDatasetData<O, S>> dataset = datasetBuilder.build(
+            environmentBuilder,
+            new EmptyContextBuilder<>(),
+            new RecommendationDatasetDataBuilder<>(),
+            trainerEnvironment
+        )) {
+            // Collect total set of objects and subjects (their identifiers).
+            Set<O> objects = dataset.compute(RecommendationDatasetData::getObjects, RecommendationTrainer::join);
+            Set<S> subjects = dataset.compute(RecommendationDatasetData::getSubjects, RecommendationTrainer::join);
+
+            // Generate initial model (object and subject matrices) initializing them with random values.
+            Map<O, Vector> objMatrix = generateRandomVectorForEach(objects, trainerEnvironment.randomNumbersGenerator());
+            Map<S, Vector> subjMatrix = generateRandomVectorForEach(subjects, trainerEnvironment.randomNumbersGenerator());
+
+            // SGD steps.
+            for (int i = 0; i < maxIterations; i++) {
```

Comment:
Why do you skip the gradient descent convergence calculation? If a gradient is converged we can stop SGD iterations.

Reply:
Correspondent task created: https://ggsystems.atlassian.net/browse/GG-22916.